### ISSUE TITLE
Get Kimi K2 Instruct full model running on multihost

### ISSUE THREAD

**Issue by jameszianxuTT** [tenstorrent/tt-xla#4678]
Currently executed via benchmarks 

Experiment logs: https://docs.google.com/spreadsheets/d/1leYUm01ye6QpedExy0Ya91IKJv64b9VbuqmVzC6wcFo/edit?gid=0#gid=0

**Comment by jameszianxuTT** [comment]
May 13 update:

Progress:
- 2 and 4 layer full model prefill benchmark tests execute E2E via benchmarks `test_kimi_k2_tp_galaxy_2_layers` -> modified to run with more layers
- These execute without crashing after overriding multihost cmd read T/O from 300->3000s, maybe needs to be increased further to 30KS (8H20M) 

Issues:
- Typecast hangs or is impossibly slow; currently bounced off host for accuracy as of https://github.com/tenstorrent/tt-mlir/pull/8140 and in test branch. 
- With random weights, we see increasing PCC degradation per layer. @2 layers 0.89, @4layers 0.53. Note -- These are both worse than the existing WH galaxy thresholds, at 0.91 -- Unclear if this model is actually functioanl
- Worker host memory spike in to_host

Estimates via linear interpolation of stats when running at 2 and 4 layers (with random weights)
- E2E execution - 265m ~ 4H
- PCC - near 0
- Random weight initialization time - 2H
- Host memory on controller (python + worker): 4T + 2T = 6TB 
- Host memory on worker: 2TB

Also, there is a huge spike in worker RSS towards the end of the prefill test in 2 layer model of 200GB. Maybe it is backtransfer of logits? 200GB / 64 devices / 32 batch = 97.5MB tensor, seems possible

<img width="750" height="336" alt="Image" src="https://github.com/user-attachments/assets/58da4b21-af89-4dd7-bd2a-9231bb7fde65" />

**Comment by jameszianxuTT** [comment]
From May 14 tests - rebaselined against May 14 tt-xla and pulled in fixes from @gengelageTT for PCC stability and @sshonTT for python proc memory efficiency, allows us to make some better predictions about how this will scale. 

Requested swap here - https://github.com/tenstorrent/exabox-infra/issues/1072 with projections on how much below. We might be able to reduce this requirement from the python side, which is currently dominant, but requires more debug to figure out where the allocations are coming from - I think @sshonTT 's patch to release temporary xla data is not active quite yet.

<img width="1037" height="640" alt="Image" src="https://github.com/user-attachments/assets/655d014c-952e-471c-a06a-74eab9e72315" />

**Comment by jameszianxuTT** [comment]
## May 15 update:

### 1. Real weights and quantization caching

Integrated @btsoiTT 's post-requantization of HF weight cache and the DS4 meta skeleton + manual weight filling into jzx/kimi in tt-forge-models, showing a nice speedup in execution on a cache hit. This saves us a significant (probably several hours - orders of magnitude?) amount of time requantizing hundreds of GB of FP8E4M3 weights into BF16, since the HF checkpoints are in an incompatible DF for us. 
- Q - Can we use FP8E4M3 directly? Probably not through PJRT / Runtime datatypes, though maybe there's a way we can smuggle it through another 8bit dtype
- Q - Can we use a bootleg reupload for kimi that saves us from requantizing stuff and maintaining our own separate cache? -- TBD testing

Quantization cache memory profiles are here and show a runtime speedup + memory overhead reduction with no accuracy degradation when using real weights.
<img width="1670" height="494" alt="Image" src="https://github.com/user-attachments/assets/46cdf52d-2a2e-40d2-817d-85c14fba9e3b" />

### 2. Decode using DistributedRMSNorm is actually broken on multihost

Gonna disable the DistributedRMSNorm fusion for now, and maybe permanently in multihost - https://github.com/tenstorrent/tt-metal/issues/44533 - this is the first observed op that does not work on multihost. In general, ttnn op implementation is multihost-capable, but there are some antipatterns in op implementation that will break multihost functionality, especially for CCL-like ops (this one is an RMS norm + all gather). 

It's very lucky this op failure doesn't manifest as runtime corruption....

Disabling this fusion [[commit](https://github.com/tenstorrent/tt-mlir/commit/4c88e05cdd824704fa532f1f38aa27e410370472)] works to get decode and runs on 2/4 layers show > 0.999 PCC w/ real weights which is great and much better than prefill. Also is corroborated by @gengelageTT 's testing on single WH which shows that decode is consistently much more accurate than prefill.

See [spreadsheet for first functional decode testing](https://docs.google.com/spreadsheets/d/1leYUm01ye6QpedExy0Ya91IKJv64b9VbuqmVzC6wcFo/edit?gid=0#gid=0&range=A34:F36)

### 3. Continued prefill accuracy issues, even with real weights

Here are some results that are hard to reason about:
- 2 layers / b32 - PCC 1.00 
- 2 layers / b64 - PCC 0.59
- 4 layers / b32 - PCC 0.094
- 4 layers / b64 - PCC 0.196

Q - Are these multihost specific degradations?
A - Not sure, something seems to have changed since rebasing, such that forcing Fabric 1D and mesh shape 4x8 init on exabox bh galaxy no longer works to enable single galaxy usage. This should be investigated further as a separate item, but without CI we are screwed in the long term.

### 4. Larger model testing

Need a lot of swap => https://tenstorrent.atlassian.net/browse/MINFRA-554. I don't think swap can be mounted on an NFS, so this needs sysadmin support, waiting for available sysadmin resources.

**Comment by jameszianxuTT** [comment]
## May 20 update

## 1. Use of bootleg bf16 model from [unslothAI's checkpoint](https://huggingface.co/unsloth/Kimi-K2-Instruct-BF16) does work and produce identical results / PCC to the original model where we manually dequantize at runtime. 

- This saves us a lot of headache maintaining a dequantize cache
- However, their checkpoint is broken and doesn't load by default, it seems unmaintained and won't work with latest transformers. So we need to load the tokenizer config from the original moonshotAI repo. 
- Using /mnt/models/kimi-forge and some modifications to weight loader to use a HF repo downloaded manually, via `hf download` API. Doesn't load manually when pointing HF_HOME, probably an artifact of how the repo is downloaded and transformers' file hierarchy expectations
  - This infra will need to be generalized further, but it is probably the way to go when using real weights in exabox cluster. This NFS share for checkpoints is 50TB so it's great for our use case. NBD - it needs to be mounted into the container! `-v /mnt/models/:/mnt/models/`

Example checkpoint log:
```
2026-05-20 18:13:59.767 | WARNING  | third_party.tt_forge_models.kimi_k2.pytorch.weight_loader:_resolve_file:88 - [weight_loader] KIMI_K2_CHECKPOINT_DIR not set, using default: /mnt/models/kimi-forge. Set KIMI_K2_CHECKPOINT_DIR to override.
```

 
## 2. Minor sidequest, why can't we use fp8 e4m3 weights as natively exists in the checkpoint for ds4 and kimi, it would save memory and time
- Pytorch supports these dataformats:
```
>>> [x for x in dir(torch) if "float8" in x]
['float8_e4m3fn', 'float8_e4m3fnuz', 'float8_e5m2', 'float8_e5m2fnuz', 'float8_e8m0fnu']
>>> [x for x in dir(torch) if "float4" in x]
['float4_e2m1fn_x2']
```
- PJRT also supports e4m3 - but runtime does not - `PJRT data type: PJRT_Buffer_Type_F8E4M3FN does not have runtime data type equivalent`

We cannot really add runtime support here. The goal here would be to take a native pytorch fp8 tensor and transit it through the whole stack to create input TTNN tensors in the same fp8 format without any casting. However, TTNN tensors do not even have a fp8 dtype, so we would basically need to add a typecast at the creation boundary which (1) wastes possibly enormous amounts time and memory to do something that we are already doing competently in python with dequantize weight cache, (2) gives up our memory savings anyways once we upcast, and (3) may break our ability to share the memory of these tensors, as we can currently only borrow from tensors of supported datatypes.

@sshonTT brings up an interesting point - TTNN exposes [dtype casting APIs](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn-nanobind/core.cpp#L181-L182), and we could exploit those to rewrite the checkpoints. PJRT then won't support the dtype, but maybe we can reinterpret it under the guise of another supported datatype, like int8.

I think this is still worth exploring later, as many models we get from HF are now seemingly expressed with these more fp8 and fp4 datatypes and having a bunch of dequantize logic in test infra due to lack of dataformat support at TTNN APIs feels like a pain point.

## 3. Two major scaling overheads - expert stacking and weight transfer
### 3a **expert stacking** - for sparse MLP, we torch.stack expert MLP weights so they can be sharded with existing sharding mechanism. This has two issues - 

(1) memory cost, we retain the stacked and copies resident in RAM, it could be evicted but it has implications for CPU golden evaluation, which would either need to unpack stacked weights, or have golden results saved for offline comparison later -- BTW - we should check that Kimi modified model actually works running on pure CPU! this will probably be insanely slow due to swap thrashing, but we will see.

 We can model expected memory pretty easily - input dim 7168 x moe MLP dim 2048 x 3 weights per expert x 384 experts per layer x 61 layers x 2 bytes per param in bf16 x 2 for doubled storage -> for a single layer, we use ~19 GB of memory which roughly lines up with this calculation.

<img width="1108" height="460" alt="Image" src="https://github.com/user-attachments/assets/7a0fafd4-0a06-4945-82d6-2660741eeb46" />

(2)  runtime cost, executing torch.stack over 1.8TB of expert weights in bf16 is going to be _very_ slow, as it involves a lot of data movement of CPU tensor data in swap memory.

Fixes: (For memory) dtype conversion is hard from the previous section, and the easiest lift would be to evict host weights and store a CPU golden offline. (For time) disk-cache the stacked experts, add complexity to model loader to handle it.

### 3b. weight transfer time

Transfer from controller to workers is slow. The link being used is theoretically 100 gigabit but it seems to take 14 minutes  to transfer 80GB of weights to each worker (160GB total, but half is on the same host so using loopback link)

<img width="527" height="562" alt="Image" src="https://github.com/user-attachments/assets/adeca1bd-acee-45e2-ab56-652c850487ea" />

WIP - https://github.com/tenstorrent/tt-mlir/commit/34539438256fdbf40b18a6a0879e463feee65573 CreateOwnedHostTensorWithDiskCache to intercept the first transfer over the wire with a request to load from a disk cache first. We can use buffer instance UIDs + a user generated string to key the cache, which will be stable between model prefill and decode but may collide between different models.

## 4. Infra Issues - Malfunctioning SSDs on at least 2 hosts?

7TB swap drives were added on Tuesday to all hosts in 110-C12 forge exabox quad, thanks to Vincent Du on metal infra. However, these swap drives (and SSDs on some exabox galaxies) have stability issues. https://tenstorrent.slack.com/archives/C08SN9MDAS3/p1779282792132229

It is unclear if we can actually write to swap without causing SSD write errors, and extended model tests on May 20 were able to softlock two galaxies requiring a reboot. This may need more help from metal infra, investigating.

update: swap disk error is reproducible - escalating for more infra help - https://tenstorrent.atlassian.net/browse/DCAMP-554

## 5. Single device hack is broken

After rebasing, it seems that singledevice execution in the exabox blackhole galaxy breaks in a different place with the old workarounds for fabric init bug; now failing later at runtime with fabric core hang. Not sure if this is bad state on one machine, workload dependent or something else random, as Sungjoon reports his machine is able to execute with the workaround

**Comment by jameszianxuTT** [comment]
Instead of logging, I use heaptrack tool to instrument large worker allocations and asked claude to look at the output -- gonna look further

Here is the crazy allocation that we're trying to suppress. 
<img width="637" height="288" alt="Image" src="https://github.com/user-attachments/assets/88e15898-bed4-4dd3-ba12-2ff615d7a7c8" />

Raw heaptrack result, this is a good tool and it is pretty interpretable - https://github.com/jameszianxuTT/logdump/blob/main/kimi_1l_heaptrack_2026-05-22_15-42-53.log

This is due to two things
- padding 64x1x163840 tensor to 64x32x163840
- casting logits to fp32 via .float() in model impl
- then all gathering it from all local devices in worker, leading to another x32 allocation

After fixing (2) and reducing batch the spike can be limited to 10GB

**Comment by jameszianxuTT** [comment]
## June 1 Update 

Last week, SSDs on exabox were determined to be too unstable and we are not able to use them for swap space without crashing the host. This means we have a hard constraint of ~550GB of host memory to run Kimi. 

There are 2 options I can think of at this point:
1. Adapting streaming to multihost 
2. Severing the runtime from torch graph tracing to create a more barebones, bespoke and memory-bounded runtime implementation, and integrating with offline CPU-only compile on a non memory bounded host. Also requires consteval pre-evaluation and disk caching
3. Use a remote controller w/o worker backing to do graph trace and dispatch on swap, while exabox workers can survive with their limited host memory requirement guarantees.

Last week's work was mainly to push (1) -- Streaming requires several conditions to work, which are validated on singlehost in Sungjoon's existing DS4 pro streaming infra. 

Some things were fixed to meet streaming requirements and general memory overhead reductions.

1. In single and multihost path, torch-xla wastes host memory by caching a tensor_data host pointer to a host-side copy in SPMD mode if a tensor is implicitly replicated (i.e. mark_sharding is not called). This can be patched with [torch-xla](https://github.com/tenstorrent/pytorch-xla/pull/36) change or more easily calling mark_sharding on _all_ inputs in tt-xla test infra 
2. In multihost, the toLayout path in PJRT impl used implicit refcount-based destruction to drop the host tensor after transferring it to device. Implicit refcount-based destruction currently does not work in multihost; as the controller will destruct its tensor but the worker not know to do so; and therefore leak. This can be fixed by force-deallocating the tensor at the implicit-drop callsite
3. Sparse MLP expert stacking saves a CPU original after stacking weights; this can be evicted if we don't do a comparison, or cache the CPU reference to avoid calculating it at runtime
4. From [above comment](https://github.com/tenstorrent/tt-xla/issues/4678#issuecomment-4520200621) - output to_host spikes due to a 'double' all gather, tilization overhead and fp32 cast. This can be reduced significantly by removing an excess `float()` cast and reducing batch from 64->16, but we still expect an at least 32x overhead of output logit actual size
5. For streaming, consteval inputs must be pushed to device memory instead of host memory; especially since all exp weights are reqquantized in consteval graphs and make up bulk of host memory. TBD - these inputs must be evicted from device memory after last user; as Sungjoon has thoroughly studied [in this comment](https://github.com/tenstorrent/tt-xla/issues/4592#issuecomment-4557469012)

Below graphs show a progression with various optimizations to the current state where:
- controller memory is bounded by model size + overhead from torch.stack; which could be further reduced by using an expert stacking cache so it just is the actual model size; of about 2TB in BF16  
- worker memory is bounded a small constant (unk. what this is, largest single host tensor maybe? -- tbd) but remains bounded below a constant 12GB which remains true even after scaling layer count. This is quite a good situation
- Device memory has not been studied but likely accumulates since CE inputs and outputs are stored on device

<img width="1730" height="468" alt="Image" src="https://github.com/user-attachments/assets/1b3bdc9b-5acc-46c5-a150-780b46beede3" />

**Comment by jameszianxuTT** [comment]
## Another June 1 update

With streaming, the memory dynamics are as expected on multihost system and a vast improvement over the original full-model requirement.

Between 2 -> 32 layers, host memory is now bounded by about 100GB on controller for single-layer loading and 12GB on worker  (with slow creep) and permits execution of **32** layers in 45 mins, which is quite good.

<img width="467" height="403" alt="Image" src="https://github.com/user-attachments/assets/7ca5929c-b801-4db3-b5b6-6be95820554d" />

Decode-only outputs look a bit degenerate but hard to say with 32 layers. PCC check will be brought up afterwards (maybe with static checkpoint or a streaming adapter for CPU-only execution.

```
2026-06-01 19:06:57.711 | INFO     | kimi_k2_streaming_test:test_kimi_k2_streaming_decode:459 -   User 0: 'The' -> ' the the the the drive drive drive drive drive drive drive drive drive drive drive drive'
```

Also metal device memory shm reading is interesting and can provide a coarse view of runtime memory dynamics with rough correlation to runtime mechanisms, showing the path that data takes from controller host to worker host, then to worker device.

<img width="515" height="569" alt="Image" src="https://github.com/user-attachments/assets/0ed33489-a2ed-425a-a1f8-e5af90ccb9e3" />

**Comment by jameszianxuTT** [comment]
## June 2 update

With streaming we can get full model execution with bounded memory on controller host, worker host and device with plenty of headroom. However outputs are degenerate for single-token decode with all tested inputs producing `&` as the next token

<img width="1049" height="603" alt="Image" src="https://github.com/user-attachments/assets/48ea636f-4a40-432b-88d8-61070b6e16ed" />

Adding in autoregressive generation:

<img width="891" height="510" alt="Image" src="https://github.com/user-attachments/assets/a86ce758-6c2b-44e4-91f2-28780414b0e5" />

Also, a 3 way comparison between:
- basic device test 2L
- streaming device test 2L
- CPU-only test

Shows that for batch 16, the results between all 3 are _very_ close


Streaming 2L (`Hello` prompt only -- note that first 4 exactly match CPU)

<img width="1760" height="704" alt="Image" src="https://github.com/user-attachments/assets/368b7a9d-c1d9-41c1-8dd5-3ed9b738d9d1" />

Device 2L

<img width="2722" height="460" alt="Image" src="https://github.com/user-attachments/assets/c0678e22-6ed5-4e88-ad47-9c625cbda2c7" />

CPU 2L

<img width="2012" height="304" alt="Image" src="https://github.com/user-attachments/assets/598470ec-2bb9-4883-a74f-100e4c8a4cd3" />
